rust/ql/lib/codeql/rust/frameworks/RustCrypto.qll

```diff
@@ -0,0 +1,57 @@
+/**
+ * Provides modeling for the `RustCrypto` family of crates (`cipher`, `digest` etc).
+ */
+
+private import rust
+private import codeql.rust.Concepts
+private import codeql.rust.dataflow.DataFlow
+
+bindingset[algorithmName]
+private string simplifyAlgorithmName(string algorithmName) {
+  // the cipher library gives triple-DES names like "TdesEee2" and "TdesEde2"
+  if algorithmName.matches("Tdes%") then result = "3des" else result = algorithmName
+}
+
+/**
+ * An operation that initializes a cipher through the `cipher::KeyInit` or
+ * `cipher::KeyIvInit` trait, for example `Des::new` or `cbc::Encryptor<des::Des>::new`.
+ */
+class StreamCipherInit extends Cryptography::CryptographicOperation::Range, DataFlow::Node {
+  string algorithmName;
+
+  StreamCipherInit() {
+    // a call to `cipher::KeyInit::new`, `cipher::KeyInit::new_from_slice`,
+    // `cipher::KeyIvInit::new`, `cipher::KeyIvInit::new_from_slices` or `rc2::Rc2::new_with_eff_key_len`.
+    exists(PathExpr p, string rawAlgorithmName |
+      this.asExpr().getExpr().(CallExpr).getFunction() = p and
+      p.getResolvedCrateOrigin().matches("%/RustCrypto%") and
+      p.getPath().getPart().getNameRef().getText() =
```

Comment:
This part is duplicated below, so I would introduce an existential variable to avoid duplication.

Reply:
Yep, this code is pretty clunky, but it's not an exact duplicate - this one is `p.getPath().getPart()` whereas below is `p.getPath().getQualifier()`.

The problem is really that we need helper predicates to make this sort of code far more concise in the first place.  There's an issue tracking this, but we haven't yet decided how best to solve it.  Opinions / discussion very welcome!

Reply:
Sorry, I missed the `getQualifier()` bit.